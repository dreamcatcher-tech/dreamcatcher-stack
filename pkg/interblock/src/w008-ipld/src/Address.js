import assert from 'assert-fast'
import { encode, prepare } from '@ipld/dag-pb'
import { CID } from 'multiformats/cid'
import { Block } from 'multiformats/block'
import { IpldInterface } from './IpldInterface'
import { sha256 } from 'multiformats/hashes/sha2'
import { Pulse } from '.'

export const cidV0FromString = (string) => {
  assert.strictEqual(typeof string, 'string')
  const bytes = encode(prepare(string))
  const hash = sha256.digest(bytes)
  const cid = CID.createV0(hash)
  return cid
}
export const generateConstantCids = () => {
  const UNKNOWN = cidV0FromString('UNKNOWN')
  const ROOT = cidV0FromString('ROOT')
  const LOOPBACK = cidV0FromString('LOOPBACK')
  const INVALID = cidV0FromString('INVALID')
  const GENESIS = cidV0FromString('GENESIS')
  const IO = cidV0FromString('IO')
  console.log({ UNKNOWN, ROOT, LOOPBACK, INVALID, GENESIS, IO })
}
// generateConstantCids()

const UNKNOWN = CID.parse('QmVQEFQi81SSbZaezrm78p333EGeYzfEZyvCnX848KaMCw')
const ROOT = CID.parse('QmSawJHmTNpaUjDYWCE9RgoHTpKbi6JD7TuGESFbtZ4ZLc')
const LOOPBACK = CID.parse('Qme2gGBx8EnSrXc5shQF867KPQd4jwubNov67KEKZbo4p3')
const INVALID = CID.parse('QmYSWwmJ4w1pZ6igGRNKcVHpBU68iaumYEjsdbMpxfAQaj')
const GENESIS = CID.parse('QmZTKF2kuFHy8isKWXpNeNa5zjeJwsHUbPbTNF1fS8HkpB')
const IO = CID.parse('QmapFBxqMFEFxSqGE45yVYKid471NoegxSriH1GpCMfUa6')
const defines = [UNKNOWN, ROOT, LOOPBACK, INVALID, GENESIS, IO]

const addressBlock = (cidV1) => {
  assert(cidV1 instanceof CID)
  assert.strictEqual(cidV1.version, 1)
  const value = prepare({ Links: [cidV1] })
  const bytes = encode(value)
  const hash = sha256.digest(bytes)
  const cid = CID.createV0(hash)
  assert.strictEqual(cid.version, 0)
  return new Block({ cid, bytes, value })
}

export class Address extends IpldInterface {
  #cid
  #ipldBlock
  static createCI(string) {
    assert.strictEqual(typeof string, 'string')
    assert(string)
    const value = prepare({ Data: string })
    const bytes = encode(value)
    const hash = sha256.digest(bytes)
    const cid = CID.createV0(hash)
    assert.strictEqual(cid.version, 0)
    const block = new Block({ cid, bytes, value })
    return Address.#createFromBlock(block)
  }
  static createUnknown() {
    return this.#createPredefined(UNKNOWN)
  }
  static createRoot() {
    return this.#createPredefined(ROOT)
  }
  static createLoopback() {
    return this.#createPredefined(LOOPBACK)
  }
  static createInvalid() {
    return this.#createPredefined(INVALID)
  }
  static createGenesis() {
    return this.#createPredefined(GENESIS)
  }
  static createIo() {
    return this.#createPredefined(IO)
  }
  static generate(pulse) {
    assert(pulse instanceof Pulse)
    assert(!pulse.isModified(), `Pulse must be crushed already`)
    assert(pulse.isGenesis(), `Pulse must be genesis`)
    // TODO check the pulse is genesis
    const block = addressBlock(pulse.cid)
    return this.#createFromBlock(block)
  }
  static #createFromBlock(block) {
    assert(block instanceof Block)
    const instance = new this()
    instance.#setCid(block.cid)
    instance.#ipldBlock = block
    return instance
  }
  static #createPredefined(cid) {
    assert(defines.includes(cid), `not predefined`)
    const instance = new this()
    instance.#setCid(cid)
    return instance
  }
  #setCid(cid) {
    assert(cid instanceof CID, `cid must be a CID, got ${cid}`)
    assert(cid.version === 0)
    this.#cid = cid
  }
  static async uncrush(rootCid, resolver, options) {
    assert(rootCid instanceof CID)
    assert.strictEqual(typeof resolver, 'function')
    for (const define of defines) {
      if (define.equals(rootCid)) {
        return this.#createPredefined(define)
      }
    }
    const block = await resolver(rootCid)
    assert(block instanceof Block)
    const instance = this.#createFromBlock(block)
    instance.#ipldBlock = undefined
    return instance
  }
  isModified() {
    return false
  }
  get ipldBlock() {
    return this.#ipldBlock
  }
  get cid() {
    return this.#cid
  }
  [Symbol.for('nodejs.util.inspect.custom')]() {
    return this.toString()
  }
  toString() {
    return `Address(${this.cid.toString().substring(0, 14)})`
  }
  crush() {
    const instance = new Address()
    instance.#setCid(this.#cid)
    return instance
  }
  getDiffBlocks() {
    const map = new Map()
    if (this.ipldBlock) {
      map.set(this.cid, this.ipldBlock)
    }
    return map
  }
  equals(address) {
    assert(address instanceof Address)
    return this.cid.equals(address.cid)
  }
  getChainId() {
    if (this.#cid === LOOPBACK) {
      return 'LOOPBACK'
    }
    if (this.#cid === ROOT) {
      return 'ROOT'
    }
    if (!this.isResolved()) {
      throw new Error(`Address not resolved: ${this.#cid}`)
    }
    return this.#cid.toString()
  }
  // TODO broaden isResolved to cover root and loopback cases
  isRoot() {
    return this.#cid === ROOT
  }
  isLoopback() {
    return this.#cid === LOOPBACK
  }
  isGenesis() {
    return this.#cid === GENESIS
  }
  isUnknown() {
    return this.#cid === UNKNOWN
  }
  isInvalid() {
    return this.#cid === INVALID
  }
  isResolved() {
    return !this.isUnknown() && !this.isInvalid() && !this.isGenesis()
  }
  isRemote() {
    return this.isResolved() && !this.isRoot() && !this.isLoopback()
  }
}
