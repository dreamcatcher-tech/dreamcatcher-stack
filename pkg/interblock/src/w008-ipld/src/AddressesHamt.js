import { Hamt } from './Hamt'
import { Address } from '.'
import assert from 'assert-fast'
export class AddressesHamt extends Hamt {
  get isBakeSkippable() {
    return true
  }
  async set(address, channelId) {
    assert(address instanceof Address)
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(address.isRemote())
    const key = address.getChainId()
    return await super.set(key, channelId)
  }
  async has(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    return await super.has(address.getChainId())
  }
  async delete(address) {
    assert(await this.has(address), `address not present ${address}`)
    return await super.delete(address.getChainId())
  }
  async get(address) {
    assert(address instanceof Address)
    return await super.get(address.getChainId())
  }
  async compare(other) {
    let { added, deleted, modified } = await super.compare(other)
    added = new Set([...added].map((key) => Address.fromChainId(key)))
    deleted = new Set([...deleted].map((key) => Address.fromChainId(key)))
    modified = new Set([...modified].map((key) => Address.fromChainId(key)))
    return { added, deleted, modified }
  }
}
