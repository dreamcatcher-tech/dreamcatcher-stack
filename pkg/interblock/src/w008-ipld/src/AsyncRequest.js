import assert from 'assert-fast'
import { Request, RequestId, Reply } from '.'
import { IpldStruct } from './IpldStruct'
import equals from 'fast-deep-equal'

/**
    type AsyncRequest struct {
        request &Request
        to String
        id RequestId
        settled optional &Reply
    }
*/
export class AsyncRequest extends IpldStruct {
  static create(request, to) {
    assert(request instanceof Request)
    assert.strictEqual(typeof to, 'string')
    assert(to)
    const instance = new AsyncRequest()
    instance.request = request
    instance.to = to
    return instance
  }
  setId(id) {
    assert(id instanceof RequestId)
    return this.setMap({ id })
  }
  settle(reply) {
    assert(reply instanceof Reply)
    assert(!this.isSettled())
    return this.setMap({ settled: reply })
  }
  isRequestMatch(request) {
    assert(request instanceof Request)
    return equals(this.request, request)
  }
  isSettled() {
    return this.settled !== undefined
  }
  assertLogic() {
    if (this.settled) {
      assert(this.id !== undefined, `cannot settle unidentified requests`)
    }
  }
}
