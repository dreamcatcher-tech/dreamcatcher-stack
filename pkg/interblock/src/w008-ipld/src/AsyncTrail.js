import assert from 'assert-fast'
import { IpldStruct } from './IpldStruct'
import {
  Pulse,
  RxReply,
  AsyncRequest,
  Reply,
  Request,
  RxRequest,
  RequestId,
} from '.'
/**
 * Lifecycle increases thru five states:
 * 1. origin      - before any reduction has taken place
 * 2. reduced     - having txs with no requestIds attached
 * 3. transmitted - all txs have requestIds attached
 * 4. fulfilled   - all txs have been settled with replies
 * 5. settled     - the trail has replied to the origin
 * 
 * 
type RequestId struct {
    channelId Int
    stream String
    requestIndex Int
}
type AsyncRequest struct {
    request &Request
    to String
    id RequestId
    reply optional &Reply
}
type RxRequest struct {
    request &Request
    requestId RequestId
}
type AsyncTrail struct {
    origin RxRequest
    settles [AsyncRequest]
    txs [AsyncRequest]
    rxReply optional &RxReply 
}
 */
export class AsyncTrail extends IpldStruct {
  settles = []
  txs = []
  static createCI() {
    const requestId = RequestId.createCI()
    const request = Request.create('TEST')
    const rxRequest = RxRequest.create(request, requestId)
    return this.create(rxRequest)
  }
  static create(origin) {
    assert(origin instanceof RxRequest)
    const instance = new this()
    Object.assign(instance, { origin })
    return instance
  }
  static createWithPulse(origin, pulse) {
    /** used when we need the response pattern, but want to make
     * one time only synchronized changes to a state variable.
     * If this mode is used, settles are disallowed, and txs
     * are disallowed, as the prescence of these items means
     * repeated execution will occur, which is counter to purpose
     * of this function.
     */
    assert(origin instanceof RxRequest)
    assert(pulse instanceof Pulse)
    return super.clone({ origin, pulse })
  }
  isPending() {
    return !this.rxReply
  }
  assertLogic() {
    assert(!this.pulse)
    assert(!this.txs.every((tx) => tx.isSettled()))
    assert(this.settles.every((tx) => tx.isSettled()))
    assert(!this.rxReply, `can only crush promised trails`)
  }
  isFulfilled() {
    // check that all settles are in fact settled
    return !this.txs.length && this.settles.every((s) => s.isSettled())
  }
  setTxs(txs) {
    assert(Array.isArray(txs))
    assert(txs.every((tx) => tx instanceof AsyncRequest))
    assert(txs.every((tx) => !tx.isSettled()))
    return this.setMap({ txs })
  }
  updateTxs(txs) {
    // used for adding requestIds to txs
    assert(Array.isArray(txs))
    assert(this.txs.length)
    assert.strictEqual(this.txs.length, txs.length)
    // TODO check requests match
    return this.setTxs(txs)
  }
  settleOrigin(reply) {
    assert(reply instanceof Reply)
    assert(!this.rxReply)
    const rxReply = RxReply.create(reply, this.origin.requestId)
    return this.setMap({ rxReply })
  }
  settleTx(rxReply) {
    assert(rxReply instanceof RxReply)
    let isMatched = false
    let txs = this.txs.map((tx) => {
      if (tx.isIdMatch(rxReply)) {
        isMatched = true
        return tx.settle(rxReply)
      }
      return tx
    })
    assert(isMatched, `no match found for ${rxReply.requestId}`)
    let { settles } = this
    if (txs.every((tx) => tx.isSettled())) {
      settles = settles.concat(txs)
      txs = []
    }
    return this.setMap({ txs, settles })
  }
  getSettles() {
    return [...this.settles]
  }
  getRequestObject() {
    return this.origin.getRequestObject()
  }
  setError(txs, error) {
    const reply = Reply.createError(error)
    return this.setTxs(txs).settleOrigin(reply)
  }
  getError() {
    assert(this.rxReply)
    assert(this.rxReply.reply.isRejection())
    return this.rxReply.reply.getRejectionError()
  }
  result() {
    if (this.isPending()) {
      return
    }
    const { reply } = this.rxReply
    if (reply.isResolve()) {
      return reply.payload
    } else {
      throw reply.getRejectionError()
    }
  }
  isTransmitted() {
    assert(!this.pulse)
    assert(this.settles.every((tx) => tx.isSettled()))
    return this.txs.every((tx) => !!tx.requestId)
  }
}
