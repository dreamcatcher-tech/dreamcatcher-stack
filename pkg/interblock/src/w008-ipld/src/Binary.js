import assert from 'assert-fast'
import { CID } from 'multiformats/cid'
import { encode, Block } from 'multiformats/block'
import { IpldInterface } from './IpldInterface'
import * as codec from 'multiformats/codecs/raw'
import { sha256 as hasher } from 'multiformats/hashes/sha2'

export class Binary extends IpldInterface {
  // TODO use the unixFS API so these objects are browseable
  #ipldBlock
  static async create(value) {
    assert(value instanceof Uint8Array)
    // TODO move to factory
    const ipldBlock = await encode({ value, codec, hasher })
    assert.strictEqual(ipldBlock.value, value)
    return this.#create(ipldBlock)
  }
  static #create(ipldBlock) {
    assert(ipldBlock instanceof Block, `#create requires a Block`)
    assert.strictEqual(ipldBlock.value, ipldBlock.bytes)
    const instance = new this()
    instance.#ipldBlock = ipldBlock
    instance.deepFreeze()
    IpldInterface.deepFreeze(ipldBlock)
    return instance
  }
  static async uncrush(rootCid, resolver, options) {
    assert(rootCid instanceof CID, `rootCid must be a CID, got ${rootCid}`)
    assert(typeof resolver === 'function', `resolver must be a function`)
    const ipldBlock = await resolver(rootCid)
    return this.#create(ipldBlock)
  }
  isModified() {
    // TODO verify that the Uint8Array cannot be altered
    return false
  }
  get ipldBlock() {
    return this.#ipldBlock
  }
  crush() {
    return this
  }
  getDiffBlocks(from) {
    if (this === from) {
      return new Map()
    }
    const blocks = new Map()
    blocks.set(this.ipldBlock.cid.toString(), this.ipldBlock)
    return blocks
  }
}
