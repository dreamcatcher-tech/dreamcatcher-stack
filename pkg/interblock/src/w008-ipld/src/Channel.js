import assert from 'assert-fast'
import {
  RxReply,
  RequestId,
  Network,
  Reply,
  Address,
  Rx,
  Tx,
  Request,
  Interpulse,
} from '.'
import Debug from 'debug'
import { IpldStruct } from './IpldStruct'
import { RxRequest } from './RxRequest'
const debug = Debug('interblock:models:channel')

/**
## Channel

`tip` matches up with precedent on the other side.

Tx and Rx are split out so that they can be stored on the block separately.
Tx needs to be separate so that remote fetching does not expose the channelId.
Rx needs to be separate to allow the ingestion of messages to be halted at any point.
Both are separate from Channel so that the potentially large HAMT that stores
all the Channel instances is updated as little as possible, whilst providing rapid
lookup to get channels that are active.

Tx needs to be hashed as it is an independent transmission, but Rx does not
need to be hashed.

The structure implements the design goal of making the Pulse be the context
of the state machine that processes all the actions.

Channel stores rx and tx only after all the activity has been wrung out of them.

```sh
type Channel struct {
    address Address                 # The remote chainId
    tx &Tx
    rx Rx
    aliases [String]                # reverse lookup
}
```
 */

export class Channel extends IpldStruct {
  static cidLinks = ['tx', 'address']
  static classMap = { tx: Tx, rx: Rx }
  static create(channelId, address = Address.createUnknown()) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(address instanceof Address)
    if (address.isRoot()) {
      assert.strictEqual(channelId, Network.FIXED_IDS.PARENT, `root not parent`)
    }

    const rx = Rx.create()
    const tx = Tx.create()
    return super.clone({ address, rx, tx, aliases: [], channelId })
  }
  static createRoot() {
    const root = Address.createRoot()
    return Channel.create(Network.FIXED_IDS.PARENT, root)
  }
  static createIo() {
    const address = Address.createIo()
    return Channel.create(Network.FIXED_IDS.IO, address)
    // TODO stop Io from being aliased
  }
  static createLoopback() {
    const address = Address.createLoopback()
    return Channel.create(Network.FIXED_IDS.LOOPBACK, address)
  }
  // TODO reject any attempts to alias loopback
  resolve(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    assert(this.isUnknown(), `Can only resolve unknown channels`)
    return this.setMap({ address })
  }
  isUnknown() {
    return this.address.isUnknown()
  }
  isRemote() {
    return this.address.isRemote()
  }
  assertLogic() {
    const { rx, tx, address, aliases, channelId } = this
    if (this.isUnknown()) {
      assert(rx.isEmpty(), 'Replies to Unknown are impossible')
      assert(!rx.tip)
    }
    if (rx.tip) {
      assert(!address.isLoopback())
    }
    if (address.isLoopback()) {
      assert(!tx.precedent)
      const banned = ['@@OPEN_CHILD']
      const systemRequests = tx.system.requests
      assert(systemRequests.every(({ type }) => !banned.includes(type)))
      assert(!aliases.length)
    }
    if (address.isRoot()) {
      assert.strictEqual(channelId, Network.FIXED_IDS.PARENT, 'Root not parent')
    }
  }
  isNext(channel) {
    assert(channel instanceof Channel)
    // TODO verify the logic follows
    return true
  }
  ingestInterpulse(interpulse) {
    assert(interpulse instanceof Interpulse)
    let { rx } = this
    rx = rx.addTip(interpulse)
    return this.setMap({ rx })
  }
  txRequest(request) {
    assert(request instanceof Request)
    const tx = this.tx.txRequest(request)
    return this.setMap({ tx })
  }
  txSystemReply(reply) {
    assert(reply instanceof Reply)
    const tx = this.tx.txSystemReply(reply)
    const rx = this.rx.shiftSystemRequest()
    return this.setMap({ tx, rx })
  }
  txReducerReply(reply) {
    assert(!this.address.isUnknown(), `Address is unknown`)
    assert(reply instanceof Reply)
    const tx = this.tx.txReducerReply(reply)
    const rx = this.rx.shiftReducerRequest()
    return this.setMap({ tx, rx })
  }
  shiftSystemReply() {
    const rx = this.rx.shiftSystemReply()
    return this.setMap({ rx })
  }
  shiftReducerReply() {
    const rx = this.rx.shiftReducerReply()
    return this.setMap({ rx })
  }
  settlePromise(rxReply) {
    const tx = this.tx.settlePromise(rxReply)
    return this.setMap({ tx })
  }
  rxSystemRequest() {
    const { channelId } = this
    const stream = 'system'
    return this.rx.system.rxRequest(channelId, stream)
  }
  rxSystemReply() {
    const { channelId } = this
    const stream = 'system'
    return this.rx.system.rxReply(channelId, stream)
  }
  rxReducerRequest() {
    const { channelId } = this
    const stream = 'reducer'
    return this.rx.reducer.rxRequest(channelId, stream)
  }
  rxReducerReply() {
    const { channelId } = this
    const stream = 'reducer'
    return this.rx.reducer.rxReply(channelId, stream)
  }
  rxIsEmpty() {
    return this.rx.isEmpty()
  }
  getNextRequestId(request) {
    assert(request instanceof Request)
    const { channelId } = this
    const stream = request.isSystem() ? 'system' : 'reducer'
    const requestIndex = this.tx[stream].requestsLength
    return RequestId.create(channelId, stream, requestIndex)
  }
}
