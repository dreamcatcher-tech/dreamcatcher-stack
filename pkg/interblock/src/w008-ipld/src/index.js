export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
