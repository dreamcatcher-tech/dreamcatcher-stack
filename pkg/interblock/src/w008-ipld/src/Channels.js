import { Hamt } from './Hamt'
import { AddressesHamt } from './AddressesHamt'
import { PulseLink, Channel, Address, Interpulse } from '.'
import assert from 'assert-fast'
import { IpldStruct } from './IpldStruct'

/**
type Channels struct {
    counter Int
    list HashMapRoot               # Map of channelIds to Channels
    addresses HashMapRoot          # reverse lookup of channels
    rxs [ Int ]
    txs [ Int ]
}



 */
const FIXED_CHANNEL_COUNT = 3

export class Channels extends IpldStruct {
  static classMap = {
    list: Hamt,
    addresses: AddressesHamt,
  }
  static create() {
    const isMutable = true
    const list = Hamt.create(Channel, isMutable)
    const counter = FIXED_CHANNEL_COUNT
    const addresses = AddressesHamt.create()
    const rxs = []
    const txs = []
    return super.clone({ counter, list, addresses, rxs, txs })
  }
  assertLogic() {}
  async has(channelId) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(channelId < this.counter)
    return await this.list.has(channelId)
  }
  async hasAddress(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    return await this.addresses.has(address)
  }
  async getByAddress(address) {
    assert(address instanceof Address)
    const channelId = await this.addresses.get(address)
    return await this.getChannel(channelId)
  }
  async deleteChannel(channelId) {
    await this.#assertChannelIdValid(channelId)
    const list = this.list.delete(channelId)
    return this.setMap({ list })
  }
  async getChannel(channelId) {
    await this.#assertChannelIdValid(channelId)
    const channel = await this.list.get(channelId)
    assert.strictEqual(channelId, channel.channelId)
    return channel
  }
  async #assertChannelIdValid(channelId) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(channelId < this.counter)
    if (channelId >= FIXED_CHANNEL_COUNT) {
      assert(await this.list.has(channelId), `channelId not present`)
    }
  }
  async addChannel(channel) {
    assert(channel instanceof Channel)
    assert(Number.isInteger(this.counter))
    assert(this.counter >= FIXED_CHANNEL_COUNT)
    let { counter, addresses } = this
    const channelId = counter++
    assert.strictEqual(channelId, channel.channelId)
    const list = await this.list.set(channelId, channel)
    const { address } = channel
    if (address.isRemote()) {
      addresses = await this.addresses.set(address, channelId)
    }
    const next = this.setMap({ counter, list, addresses })
    return next.updateActives(channelId, channel)
  }
  async updateChannel(channel) {
    const { channelId } = channel
    await this.#assertChannelIdValid(channelId)
    let previous
    const isPrevious = await this.list.has(channelId)
    if (channelId >= FIXED_CHANNEL_COUNT || isPrevious) {
      previous = await this.list.get(channelId)
      assert(previous.isNext(channel))
      // TODO if channel is identical, return
    }
    const list = await this.list.set(channelId, channel)

    let { addresses } = this
    const isResolved = previous && !previous.isRemote() && channel.isRemote()
    const isNewResolved = !previous && channel.isRemote()
    if (isResolved || isNewResolved) {
      const { address } = channel
      addresses = await addresses.set(address, channelId)
    }
    const next = this.setMap({ list, addresses })
    return next.updateActives(channelId, channel)
  }
  updateActives(channelId, channel) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(channel instanceof Channel)
    const { address } = channel
    let { rxs, txs } = this
    if (!channel.rxIsEmpty()) {
      assert(!address.isRoot())
      if (!rxs.includes(channelId)) {
        rxs = [...rxs, channelId]
      }
    } else {
      if (rxs.includes(channelId)) {
        rxs = rxs.filter((id) => id !== channelId)
        if (address.isInvalid()) {
          // TODO delete the channel from the list and clean up aliases
        }
      }
    }
    if (rxs !== this.rxs) {
      // ensure loopback, parent, and io come before others
      rxs.sort((a, b) => a - b)
    }
    if (!channel.tx.isEmpty()) {
      assert(!address.isRoot())
      if (!txs.includes(channelId) && !address.isIo() && address.isResolved()) {
        txs = [...txs, channelId]
      }
    } else {
      if (txs.includes(channelId)) {
        txs = txs.filter((id) => id !== channelId)
      }
    }
    return this.setMap({ rxs, txs })
  }
  async getTx(channelId) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(this.txs.includes(channelId))
    const { tx } = await this.list.get(channelId)
    assert(!tx.isEmpty())
    return tx
  }
  async blankTxs(precedent) {
    assert(precedent instanceof PulseLink)
    let { list } = this
    for (const channelId of this.txs) {
      let channel = await list.get(channelId)
      assert(!channel.tx.isEmpty())
      const tx = channel.tx.blank(precedent)
      channel = channel.setMap({ tx })
      list = await list.set(channelId, channel)
    }
    return this.setMap({ list, txs: [] })
  }
  async ingestInterpulse(interpulse) {
    assert(interpulse instanceof Interpulse)
    const { source } = interpulse
    assert(await this.hasAddress(source), `No address: ${source}`)
    let channel = await this.getByAddress(source)
    channel = channel.ingestInterpulse(interpulse)
    return await this.updateChannel(channel)
  }
  async *rxChannels() {
    for (const channelId of this.rxs) {
      const channel = await this.getChannel(channelId)
      assert(!channel.rxIsEmpty())
      yield channel
    }
  }
}
