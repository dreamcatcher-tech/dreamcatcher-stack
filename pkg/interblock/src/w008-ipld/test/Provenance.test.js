import { assert } from 'chai/index.mjs'
import Debug from 'debug'
import { Provenance, Dmz } from '..'
import { Timestamp } from '../src'
Debug.enable()

describe('provenance', () => {
  test('basic', async () => {
    const now = new Date('2022-03-25T03:21:34.127Z')
    const timestamp = Timestamp.create(now)
    const dmz = await Dmz.create({ timestamp }).crush()
    let provenance = Provenance.createGenesis(dmz)
    provenance = await provenance.crush()
    const diffs = provenance.getDiffBlocks()
    const cids = [...diffs.keys()]
    expect(cids).toMatchSnapshot()
    const values = [...diffs.values()].map((b) => b.value)
    expect(values).toMatchSnapshot()
  })
  test.todo('non genesis lineage always has at least one hash')
  describe('multi parent merge', () => {
    test.todo('only one ultimate genesis block allowed')
    test.todo('no duplicate parents')
  })
})
