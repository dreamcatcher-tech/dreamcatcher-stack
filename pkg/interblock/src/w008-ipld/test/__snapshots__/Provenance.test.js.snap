// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provenance basic 1`] = `
Array [
  "bafyreidi5qmyhv2slv5sqm2mkojrm4ki2j3cixykgfiudmktyp7jy5sauu",
  "bafyreiciavtdqojmryr4xb6ikuovrgkc5w5hau7gmlnyjzlipjizttsdfi",
  "bafyreigurnbmacoa6vokasdxnhknxmjf2jj3vzujitejowaeq3ihna5zxm",
]
`;

exports[`provenance basic 2`] = `
Array [
  Object {
    "address": Object {
      "code": 112,
      "hash": Uint8Array [
        18,
        32,
        165,
        37,
        156,
        83,
        149,
        142,
        39,
        249,
        188,
        186,
        117,
        242,
        232,
        16,
        171,
        41,
        157,
        93,
        94,
        179,
        164,
        38,
        153,
        8,
        214,
        100,
        55,
        32,
        182,
        110,
        175,
        204,
      ],
      "version": 0,
    },
    "dmz": Object {
      "code": 113,
      "hash": Uint8Array [
        18,
        32,
        7,
        173,
        102,
        172,
        13,
        107,
        49,
        228,
        125,
        170,
        87,
        213,
        125,
        251,
        182,
        82,
        200,
        203,
        121,
        7,
        87,
        168,
        83,
        33,
        148,
        222,
        89,
        232,
        190,
        188,
        91,
        33,
      ],
      "version": 1,
    },
    "lineages": Array [],
    "states": "TODO",
    "turnovers": "TODO",
    "validators": Object {
      "code": 113,
      "hash": Uint8Array [
        18,
        32,
        72,
        5,
        102,
        56,
        57,
        44,
        142,
        35,
        203,
        135,
        200,
        85,
        29,
        88,
        153,
        66,
        237,
        186,
        112,
        83,
        230,
        98,
        219,
        132,
        229,
        104,
        122,
        81,
        153,
        206,
        67,
        42,
      ],
      "version": 1,
    },
  },
  Object {
    "publicKeys": Array [
      Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          212,
          139,
          66,
          192,
          9,
          192,
          245,
          92,
          160,
          72,
          119,
          105,
          212,
          219,
          177,
          37,
          210,
          83,
          186,
          230,
          137,
          68,
          200,
          151,
          88,
          4,
          134,
          208,
          118,
          131,
          185,
          187,
        ],
        "version": 1,
      },
    ],
    "quorumThreshold": 1,
  },
  Object {
    "algorithm": "Secp256k1",
    "key": "puEKUvWNGow9H4HtyXKxEWoaERpZ9BzKV9jt2Tx8FBys",
    "name": "CI",
  },
]
`;
